num.nix

```diff
@@ -38,62 +43,105 @@ rec {
     then y
     else x;
 
-  mod = base: int: base - (int * (builtins.div base int));
+  /* compare :: number -> number -> int
 
-  /* even :: integer -> bool
+     Compares two numbers and returns -1 if the first is less than the second, 0
+     if they are equal, or 1 if the first is greater than the second.
   */
-  even = x: mod x 2 == 0;
+  compare = x: y:
+    if x < y
+      then -1
+    else if x > y
+      then 1
+    else 0;
 
-  /* odd :: integer -> bool
+  /* quot :: integer -> integer -> integer
+
+     Integer division, truncated towards 0. This is an alias for 'builtins.div'.
   */
-  odd = x: mod x 2 == 1;
+  quot = builtins.div;
 
-  /* divMod :: Integral a => a -> a -> (a, a)
+  /* rem :: integer -> integer -> integer
+
+     Integer remainder. For integers m and n, this has the property that
+     "n * (quot m n) + rem m n = m".
+  */
+  rem = base: int: base - (int * (quot base int));
+
+  /* div :: integer -> integer -> integer
+
+     Integer division, truncated towards negative infinity. Despite the name,
+     note that this is not the same as 'builtins.div', which truncates towards 0.
   */
-  divMod = n: d:
-    let qr = quotRem n d;
-        q = qr._0;
-        r = qr._1;
-    in if signum r == negate (signum d)
-       then { _0 = q - 1; _1 = r + d; }
-       else qr;
+  div = base: int:
+    let q = quot base int;
+    in if (base < 0) != (int < 0)
+      then q - 1
+      else q;
+
+  /* mod :: integer -> integer -> integer
+
+     Integer modulus. For integers m and n, this has the property that
+     "n * (div m n) + mod m n = m".
+  */
+  mod = base: int: base - (int * (div base int));
 
   /* quotRem :: Integral a => a -> a -> (a, a)
   */
-  quotRem = n: d: throw "implement quotRem";
+  quotRem = n: d: { _0 = quot n d; _1 = rem n d; };
+
+  /* divMod :: Integral a => a -> a -> (a, a)
+  */
+  divMod = n: d: { _0 = div n d; _1 = mod n d; };
 
+  /* even :: integer -> bool
+  */
+  even = x: rem x 2 == 0;
+
+  /* odd :: integer -> bool
+  */
+  odd = x: rem x 2 != 0;
+
+  /* fac :: integer -> integer
+
+     Integer factorial.
+  */
   fac = n:
     if n < 0
     then throw "std.num.fac: argument is negative"
     else (if n == 0 then 1 else n * fac (n - 1));
 
+  /* pow :: number -> integer -> number
+
+     Integer exponentiation. Note that this only handles positive integer exponents.
+  */
   pow = base0: exponent0:
     let pow' = base: exponent: value:
         if exponent == 0
-        then 1
+          then 1
         else if exponent <= 1
-             then value
-             else (pow' base (exponent - 1) (value * base));
-    in pow' base0 exponent0 base0;
+          then value
+        else pow' base (exponent - 1) (value * base);
+    in if base0 == 0 || base0 == 1 then 1 else pow' base0 exponent0 base0;
 
   pi = 3.141592653589793238;
 
-  toFloat = x: x + 0.0;
+  /* toFloat :: int -> float
```

Comment:
will technically convert any number to a float, though i guess we only really have ints and floats anyway. but still i think it would be more precise to say `number -> float` since in types.nix we make distinctions between integral types that are more fine grained than just `int`

Reply:
good catch!